### PR TITLE
Feature/1753 program var apis

### PR DESCRIPTION
**Pull request by davidnich**
this pull request features the following new Qore APIs:
* `Program::getGlobalVars()`
* `Program::setGlobalVarValue()`
* `get_global_vars()`
* `get_local_vars()`
* `set_global_var_value()`
* `set_local_var_value()`

and the following public C++ APIs:
* `QoreHashNode* QoreProgram::getGlobalVars() const`
* `int QoreProgram::setGlobalVarValue(const char* name, QoreValue val, ExceptionSink* xsink)`
* `QoreHashNode* q_get_thread_local_vars(int level, ExceptionSink* xsink)`
* `int q_thread_set_var_value(const char* name, const QoreValue& val, ExceptionSink* xsink)`

Also note that type information is not stored at runtime for local variables, therefore it's possible to assign a value to a local variable that is invalid according to the local variable's definition.  Ex: a variable declared `int x;` can then be assigned to any possible value (ex: `string`) with these new APIs, because the type information for the local variable is not available at runtime.

Additionally, only local variables in the current stack frame are assignable.

As discussed, `PO_ALLOW_DEBUGGING` and `QDOM_DEBUG_INSECURE` have been implemented to restrict functionality that could be dangerous (ex: viewing local vars in other stack frames, modifying local variables which allows setting types not allowed by the local var's declaration, etc).

### CODE REVIEW

File: include/qore/QoreLib.h
Comment (omusil24) [inline include/qore/QoreLib.h]: Wrong variable name either in the function declaration or in the docs.

**Review by omusil24 — APPROVED** [review]

**Comment by omusil24** [comment]
@davidnich Should I wait for @tmandys review?

**Comment by davidnich** [comment]
@omusil24 yes, please, this change is for him, and I would like his OK before we merge

**Comment by tmandys** [comment]
Thinking about DebugProgram, I'm uncertain about PO_ALLOW_DEBUGGING.
Currently it raises exception when a such debug function is used, right?
Is not better approach that no exception is raised but the function does
silently nothing? I would enable having debug stuff together with code
and enable just changing flag. I know it is not so trivial because of
class instances,...

The debug API could be also used for automation which I can imagine 
for testing, stressing etc. E.g. set debug callbacks to specific function,
change values. One use case is to implement probes for lock, counters and 
simulate deadlock vulnerability, i.e. what Pavel does using sleep() in Qorus now.

get_local_vars() is rather beyond debug logic because Program is to be debugged.
It corresponds to get_all_thread_stack which is per thread, i.e.
you can see local value from previous program (even probably regardless PO_ALLOW_DEBUGGING
of parent program).

Does get_local_vars(frame) get correctly vars when are overloaded as local block vars e.g. 
{int i=1; {int i=0; }} ? Respectively it's probably one frame, so should return i:0.

Anyway seems Ok.

**Review by tmandys — APPROVED** [review]
I put some notes as general pull request comment

**Pull request merged**
